{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cog-24.cog.csv\tcog-24.fun.tab\t\t  Readme.COG2024.txt\n",
      "cog-24.def.tab\tpathway_result_table.tsv\n"
     ]
    }
   ],
   "source": [
    "!ls /home/koala/my_projects/metagenome_metabolic_profiling/data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# P.1. Load COG ID to Functional Category mapping (cog-24.def.tab file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG Functional category ID</th>\n",
       "      <th>Functional group</th>\n",
       "      <th>RGB color</th>\n",
       "      <th>FC description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>J</td>\n",
       "      <td>1</td>\n",
       "      <td>FCCCFC</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>FCDCFC</td>\n",
       "      <td>RNA processing and modification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>FCDCEC</td>\n",
       "      <td>Transcription</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>FCDCDC</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "      <td>FCDCCC</td>\n",
       "      <td>Chromatin structure and dynamics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  COG Functional category ID  Functional group RGB color  \\\n",
       "0                          J                 1    FCCCFC   \n",
       "1                          A                 1    FCDCFC   \n",
       "2                          K                 1    FCDCEC   \n",
       "3                          L                 1    FCDCDC   \n",
       "4                          B                 1    FCDCCC   \n",
       "\n",
       "                                    FC description  \n",
       "0  Translation, ribosomal structure and biogenesis  \n",
       "1                  RNA processing and modification  \n",
       "2                                    Transcription  \n",
       "3            Replication, recombination and repair  \n",
       "4                 Chromatin structure and dynamics  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# descriptions of COG functional categories\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "cog_fun_df = pd.read_csv(\n",
    "    cog_fun_file, \n",
    "    sep=\"\\t\", \n",
    "    names=['COG Functional category ID', 'Functional group', 'RGB color', 'FC description']\n",
    "    )\n",
    "cog_fun_df.head()\n",
    "\n",
    "# Functional groups\n",
    "# 1:\tINFORMATION STORAGE AND PROCESSING\n",
    "# 2: \tCELLULAR PROCESSES AND SIGNALING\n",
    "# 3:\tMETABOLISM\n",
    "# 4:\tPOORLY CHARACTERIZED"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# P.2. Functional Category descriptions and groups (cog-24.fun.edited.tab file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG ID</th>\n",
       "      <th>COG Functional category ID</th>\n",
       "      <th>COG name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Pathway</th>\n",
       "      <th>PubMed ID</th>\n",
       "      <th>PDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COG0001</td>\n",
       "      <td>H</td>\n",
       "      <td>Glutamate-1-semialdehyde aminotransferase</td>\n",
       "      <td>HemL</td>\n",
       "      <td>Heme biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2CFB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COG0002</td>\n",
       "      <td>E</td>\n",
       "      <td>N-acetyl-gamma-glutamylphosphate reductase</td>\n",
       "      <td>ArgC</td>\n",
       "      <td>Arginine biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3DR3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COG0003</td>\n",
       "      <td>P</td>\n",
       "      <td>Anion-transporting ATPase, ArsA/GET3 family</td>\n",
       "      <td>ArsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1F48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COG0004</td>\n",
       "      <td>P</td>\n",
       "      <td>Ammonia channel protein AmtB</td>\n",
       "      <td>AmtB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1U77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COG0005</td>\n",
       "      <td>F</td>\n",
       "      <td>Purine nucleoside phosphorylase</td>\n",
       "      <td>XapA</td>\n",
       "      <td>Purine salvage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1YQQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    COG ID COG Functional category ID  \\\n",
       "0  COG0001                          H   \n",
       "1  COG0002                          E   \n",
       "2  COG0003                          P   \n",
       "3  COG0004                          P   \n",
       "4  COG0005                          F   \n",
       "\n",
       "                                      COG name  Gene                Pathway  \\\n",
       "0    Glutamate-1-semialdehyde aminotransferase  HemL      Heme biosynthesis   \n",
       "1   N-acetyl-gamma-glutamylphosphate reductase  ArgC  Arginine biosynthesis   \n",
       "2  Anion-transporting ATPase, ArsA/GET3 family  ArsA                    NaN   \n",
       "3                 Ammonia channel protein AmtB  AmtB                    NaN   \n",
       "4              Purine nucleoside phosphorylase  XapA         Purine salvage   \n",
       "\n",
       "  PubMed ID PDB ID  \n",
       "0       NaN   2CFB  \n",
       "1       NaN   3DR3  \n",
       "2       NaN   1F48  \n",
       "3       NaN   1U77  \n",
       "4       NaN   1YQQ  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# COG descriptions\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_def_df = pd.read_csv(\n",
    "    cog_def_file, \n",
    "    sep=\"\\t\", \n",
    "    names=['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway', 'PubMed ID', 'PDB ID']\n",
    "    )\n",
    "cog_def_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG ID</th>\n",
       "      <th>COG Functional category ID</th>\n",
       "      <th>COG name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Pathway</th>\n",
       "      <th>PubMed ID</th>\n",
       "      <th>PDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COG0001</td>\n",
       "      <td>H</td>\n",
       "      <td>Glutamate-1-semialdehyde aminotransferase</td>\n",
       "      <td>HemL</td>\n",
       "      <td>Heme biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2CFB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COG0007</td>\n",
       "      <td>H</td>\n",
       "      <td>Uroporphyrinogen-III methylase (siroheme synth...</td>\n",
       "      <td>CysG</td>\n",
       "      <td>Cobalamine/B12 biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1PJQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>COG0011</td>\n",
       "      <td>H</td>\n",
       "      <td>Thiamin-binding stress-response protein YqgV, ...</td>\n",
       "      <td>YqgV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20471400</td>\n",
       "      <td>1LXJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>COG0029</td>\n",
       "      <td>H</td>\n",
       "      <td>Aspartate oxidase</td>\n",
       "      <td>NadB</td>\n",
       "      <td>NAD biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1KNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>COG0043</td>\n",
       "      <td>H</td>\n",
       "      <td>3-polyprenyl-4-hydroxybenzoate decarboxylase</td>\n",
       "      <td>UbiD</td>\n",
       "      <td>Ubiquinone biosynthesis</td>\n",
       "      <td>28057757</td>\n",
       "      <td>2IDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4702</th>\n",
       "      <td>COG5749</td>\n",
       "      <td>H</td>\n",
       "      <td>Chlorophyllide a oxygenase/letal leaf spot pro...</td>\n",
       "      <td>PobA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14657372;22366162</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4706</th>\n",
       "      <td>COG5753</td>\n",
       "      <td>H</td>\n",
       "      <td>Bacteriochlorophyllide reductase subunit BchX</td>\n",
       "      <td>BchX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21886856</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4707</th>\n",
       "      <td>COG5754</td>\n",
       "      <td>H</td>\n",
       "      <td>Bacteriochlorophyllide reductase subunit BchY</td>\n",
       "      <td>BchY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21886856</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4708</th>\n",
       "      <td>COG5755</td>\n",
       "      <td>H</td>\n",
       "      <td>Bacteriochlorophyllide reductase subunit BchZ</td>\n",
       "      <td>BchZ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21886856</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4712</th>\n",
       "      <td>COG5759</td>\n",
       "      <td>H</td>\n",
       "      <td>Mg-protoporphyrin IX monomethyl ester cyclase ...</td>\n",
       "      <td>Ycf59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21388705</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       COG ID COG Functional category ID  \\\n",
       "0     COG0001                          H   \n",
       "6     COG0007                          H   \n",
       "10    COG0011                          H   \n",
       "28    COG0029                          H   \n",
       "41    COG0043                          H   \n",
       "...       ...                        ...   \n",
       "4702  COG5749                          H   \n",
       "4706  COG5753                          H   \n",
       "4707  COG5754                          H   \n",
       "4708  COG5755                          H   \n",
       "4712  COG5759                          H   \n",
       "\n",
       "                                               COG name   Gene  \\\n",
       "0             Glutamate-1-semialdehyde aminotransferase   HemL   \n",
       "6     Uroporphyrinogen-III methylase (siroheme synth...   CysG   \n",
       "10    Thiamin-binding stress-response protein YqgV, ...   YqgV   \n",
       "28                                    Aspartate oxidase   NadB   \n",
       "41         3-polyprenyl-4-hydroxybenzoate decarboxylase   UbiD   \n",
       "...                                                 ...    ...   \n",
       "4702  Chlorophyllide a oxygenase/letal leaf spot pro...   PobA   \n",
       "4706      Bacteriochlorophyllide reductase subunit BchX   BchX   \n",
       "4707      Bacteriochlorophyllide reductase subunit BchY   BchY   \n",
       "4708      Bacteriochlorophyllide reductase subunit BchZ   BchZ   \n",
       "4712  Mg-protoporphyrin IX monomethyl ester cyclase ...  Ycf59   \n",
       "\n",
       "                          Pathway          PubMed ID PDB ID  \n",
       "0               Heme biosynthesis                NaN   2CFB  \n",
       "6     Cobalamine/B12 biosynthesis                NaN   1PJQ  \n",
       "10                            NaN           20471400   1LXJ  \n",
       "28               NAD biosynthesis                NaN   1KNP  \n",
       "41        Ubiquinone biosynthesis           28057757   2IDB  \n",
       "...                           ...                ...    ...  \n",
       "4702                          NaN  14657372;22366162    NaN  \n",
       "4706                          NaN           21886856    NaN  \n",
       "4707                          NaN           21886856    NaN  \n",
       "4708                          NaN           21886856    NaN  \n",
       "4712                          NaN           21388705    NaN  \n",
       "\n",
       "[216 rows x 7 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_def_df[cog_def_df['COG Functional category ID'] == \"H\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Heme biosynthesis\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Purine salvage\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "Heme biosynthesis\n",
      "tRNA modification\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Proline biosynthesis\n",
      "Purine biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aminoacyl-tRNA synthetases\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "Pentose phosphate pathway\n",
      "Pyruvate oxidation\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "Purine biosynthesis\n",
      "Purine biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "NAD biosynthesis\n",
      "16S rRNA modification\n",
      "Cysteine biosynthesis\n",
      "nan\n",
      "Purine biosynthesis\n",
      "Pyrimidine salvage\n",
      "Pentose phosphate pathway\n",
      "tRNA modification\n",
      "nan\n",
      "TCA cycle\n",
      "Histidine biosynthesis\n",
      "Purine biosynthesis\n",
      "tRNA modification\n",
      "Ubiquinone biosynthesis\n",
      "Pyrimidine biosynthesis\n",
      "TCA cycle\n",
      "Purine biosynthesis\n",
      "Purine biosynthesis\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "Translation factors\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "FoF1-type ATP synthase\n",
      "FoF1-type ATP synthase\n",
      "Glycolysis\n",
      "nan\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "Glutamine biosynthesis\n",
      "nan\n",
      "Glutamine biosynthesis\n",
      "Glutamine biosynthesis\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "TCA cycle\n",
      "Serine biosynthesis\n",
      "Pantothenate/CoA biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Arginine biosynthesis\n",
      "Cobalamine/B12 biosynthesis\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Aromatic amino acid biosynthesis\n",
      "Threonine biosynthesis\n",
      "nan\n",
      "RNA polymerase\n",
      "RNA polymerase\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Lipoate biosynthesis\n",
      "Ribosome 30S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "tRNA modification\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "Purine biosynthesis\n",
      "Pyrimidine biosynthesis\n",
      "Histidine biosynthesis\n",
      "Histidine biosynthesis\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "Serine biosynthesis\n",
      "Serine biosynthesis\n",
      "Heme biosynthesis\n",
      "TCA cycle\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "23S rRNA modification\n",
      "Riboflavin/FAD biosynthesis\n",
      "Histidine biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "Pentose phosphate pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Thymidylate biosynthesis\n",
      "Glycolysis\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "tRNA modification\n",
      "Histidine biosynthesis\n",
      "Biotin biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Lysine biosynthesis\n",
      "Arginine biosynthesis\n",
      "Purine biosynthesis\n",
      "Histidine biosynthesis\n",
      "Histidine biosynthesis\n",
      "Histidine biosynthesis\n",
      "Isoprenoid biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "16S rRNA modification\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "Glycolysis\n",
      "Glycolysis\n",
      "Purine biosynthesis\n",
      "Purine biosynthesis\n",
      "Purine biosynthesis\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Cysteine biosynthesis\n",
      "Biotin biosynthesis\n",
      "NAD biosynthesis\n",
      "Gluconeogenesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Arginine biosynthesis\n",
      "Biotin biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "Arginine biosynthesis\n",
      "Glycolysis\n",
      "Pyrimidine biosynthesis\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "NAD biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aminoacyl-tRNA synthetases\n",
      "Glutamine biosynthesis\n",
      "Cysteine biosynthesis\n",
      "Pentose phosphate pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Heme biosynthesis\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "Glycolysis\n",
      "Methionine biosynthesis\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "Ribosome 50S subunit\n",
      "Sec pathway\n",
      "RNA polymerase\n",
      "Ribosome 50S subunit\n",
      "Phospholipid biosynthesis\n",
      "Glycolysis\n",
      "nan\n",
      "Thymidylate biosynthesis\n",
      "Pyrimidine salvage\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "Pyridoxal phosphate biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "FoF1-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "Fatty acid biosynthesis\n",
      "Pantothenate/CoA biosynthesis\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "Histidine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Folate biosynthesis\n",
      "Proline biosynthesis\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "16S rRNA modification\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "Pyrimidine biosynthesis\n",
      "Folate biosynthesis\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "Translation factors\n",
      "Ribosome 50S subunit\n",
      "Ribosome 50S subunit\n",
      "23S rRNA modification\n",
      "Folate biosynthesis\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Purine biosynthesis\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "Folate biosynthesis\n",
      "Molybdopterin biosynthesis\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "nan\n",
      "16S rRNA modification\n",
      "Molybdopterin biosynthesis\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "16S rRNA modification\n",
      "Lipoate biosynthesis\n",
      "Lipoate biosynthesis\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Folate biosynthesis\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Fatty acid biosynthesis\n",
      "Ribosome 50S subunit\n",
      "Glutamine biosynthesis\n",
      "Ribosome 50S subunit\n",
      "tRNA modification\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Biotin biosynthesis\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Proline biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "nan\n",
      "FoF1-type ATP synthase\n",
      "FoF1-type ATP synthase\n",
      "16S rRNA modification\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Ribosome 30S subunit\n",
      "Translation factors\n",
      "Pentose phosphate pathway\n",
      "Pentose phosphate pathway\n",
      "Pentose phosphate pathway\n",
      "nan\n",
      "nan\n",
      "Asparagine biosynthesis\n",
      "Cobalamine/B12 biosynthesis\n",
      "Cysteine biosynthesis\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "TCA cycle\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "NAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "Glycine cleavage\n",
      "Glycine cleavage\n",
      "nan\n",
      "nan\n",
      "Heme biosynthesis\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "Pyrimidine biosynthesis\n",
      "Histidine biosynthesis, Purine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "TCA cycle\n",
      "nan\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "TCA cycle\n",
      "Translation factors\n",
      "nan\n",
      "tRNA modification\n",
      "Gluconeogenesis\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glutamine biosynthesis\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "Threonine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Biotin biosynthesis\n",
      "Purine salvage\n",
      "Pyrimidine biosynthesis\n",
      "Arginine biosynthesis\n",
      "Proline degradation\n",
      "nan\n",
      "Pyruvate oxidation\n",
      "Glycine cleavage\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Purine biosynthesis\n",
      "nan\n",
      "Purine biosynthesis\n",
      "Purine biosynthesis\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "Glycolysis\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "Pyrimidine biosynthesis\n",
      "Cysteine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "TCA cycle\n",
      "Ribosome 30S subunit\n",
      "Pyrimidine biosynthesis\n",
      "Sec pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Sec pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "Serine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Pyrimidine biosynthesis\n",
      "23S rRNA modification\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "TCA cycle\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "Gluconeogenesis\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "Pyrimidine salvagetRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "tRNA modification\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Purine salvage\n",
      "Heme biosynthesis\n",
      "FoF1-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Sec pathway\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "Sec pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "Pentose phosphate pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Sec pathway\n",
      "Mureine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "FoF1-type ATP synthase\n",
      "FoF1-type ATP synthase\n",
      "NADH dehydrogenase\n",
      "Arch_secret\n",
      "nan\n",
      "Heme biosynthesis\n",
      "Thymidylate biosynthesis\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "16S rRNA modification\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thymidylate biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "Lipid A biosynthesis\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "Mureine biosynthesis\n",
      "Mureine biosynthesis\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "Lipid A biosynthesis\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "Purine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Entner-Doudoroff pathway\n",
      "Folate biosynthesis\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Tat pathway\n",
      "16S rRNA modification\n",
      "Riboflavin/FAD biosynthesis\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "Purine salvage\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "Fatty acid biosynthesis\n",
      "23S rRNA modification\n",
      "nan\n",
      "Ribosome 30S subunit\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "NADH dehydrogenase\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "Pantothenate/CoA biosynthesis\n",
      "Pyridoxal phosphate biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycine cleavage\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "NADH dehydrogenase\n",
      "NADH dehydrogenase\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "Proline degradation\n",
      "Pyruvate oxidation\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Lipid A biosynthesis\n",
      "Lipid A biosynthesis\n",
      "Cysteine biosynthesis\n",
      "nan\n",
      "Lysine biosynthesisTCA cycle\n",
      "TCA cycle\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "TCA cycle\n",
      "nan\n",
      "nan\n",
      "NAD biosynthesis\n",
      "NAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "Pyruvate oxidation\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "Translation factors\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "A/V-type ATP synthase\n",
      "A/V-type ATP synthase\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Mureine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "nan\n",
      "Translation factors\n",
      "16S rRNA modification\n",
      "nan\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "RNA polymerase\n",
      "nan\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "Lipid A biosynthesis\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type VII secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "Glycine cleavagePyruvate oxidation\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "A/V-type ATP synthase\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Gluconeogenesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Sec pathway\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thymidylate biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "Pyrimidine degradation\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "Thymidylate biosynthesis\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IVB secretion/DNA transfer system\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "Aminoacyl-tRNA synthetases\n",
      "16S rRNA modification\n",
      "nan\n",
      "Histidine biosynthesis\n",
      "nan\n",
      "nan\n",
      "A/V-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "A/V-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "Sec pathway\n",
      "nan\n",
      "Biotin biosynthesis\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "A/V-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoleucine, leucine, valine biosynthesis\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "Folate biosynthesis\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "Lipid A biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "A/V-type ATP synthase\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Folate biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Lipid A biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "23S rRNA modification\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Heme biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Lipid A biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "Arch_secret\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "Purine biosynthesis\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Folate biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "Tat pathway\n",
      "nan\n",
      "Purine biosynthesis\n",
      "Pantothenate/CoA biosynthesis\n",
      "Glycolysis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "TCA cycle\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Gluconeogenesis\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "Serine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "TCA cycle\n",
      "Sec pathway\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "nan\n",
      "Fatty acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "RNA polymerase\n",
      "Archaeal ribosomal proteins\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "Aminoacyl-tRNA synthetases\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cysteine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "TCA cycle\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Flp/Tad pili\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Menaquinone biosynthesis\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Biotin biosynthesis\n",
      "Archaeal ribosomal proteins\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "16S rRNA modification\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "Archaeal ribosomal proteins\n",
      "Glycolysis\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thiamine biosynthesis\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Ribosome 50S subunit\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "TCA cycle, glyoxylate bypass\n",
      "TCA cycle, glyoxylate bypass\n",
      "Biotin biosynthesis\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "nan\n",
      "Purine salvage\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "Pyridoxal phosphate biosynthesis\n",
      "Cobalamine/B12 biosynthesis\n",
      "Cobalamine/B12 biosynthesis\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "16S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Threonine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "Proline biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Riboflavin/FAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "Asparagine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "A/V-type ATP synthase\n",
      "nan\n",
      "16S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine salvage\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type V secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "TCA cycle\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "Methionine biosynthesis\n",
      "Archaella\n",
      "Cobalamine/B12 biosynthesis\n",
      "Aromatic amino acid biosynthesis\n",
      "Lipid A biosynthesis\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "Cysteine biosynthesis\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type V secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "nan\n",
      "nan\n",
      "Phospholipid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type VI secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "NAD biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "16S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "FoF1-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Lysine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "Archaella\n",
      "Archaella\n",
      "Archaella\n",
      "Archaella\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Purine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "T6SS\n",
      "T6SS\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "23S rRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "T6SS\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "Type IV secretion system\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Flp/Tad pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "NADH dehydrogenase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "Flp/Tad pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Gluconeogenesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Isoprenoid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type VI secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type X secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyruvate oxidation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type VI secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Molybdopterin biosynthesis\n",
      "Molybdopterin biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pyrimidine degradation\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Aromatic amino acid biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type VI secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Thymidylate biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "Na+-translocating Fd:NADH oxidoreductase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Cobalamine/B12 biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Non-phosphorylated Entner-Doudoroff pathway\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Arginine biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Flp/Tad pili\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Urea cycle\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Glycolysis\n",
      "nan\n",
      "Ubiquinone biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Pantothenate/CoA biosynthesis\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Translation factors\n",
      "Translation factors\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type IV secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type V secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "tRNA modification\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "FoF1-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Archaeal ribosomal proteins\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "RNA polymerase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type V secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Type III secretion system\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem I\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "Photosystem II\n",
      "nan\n",
      "Photosystem I\n",
      "nan\n",
      "Photosystem I\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "FoF1-type ATP synthase\n",
      "FoF1-type ATP synthase\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "nan\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "CRISPR-Cas system\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Type II secretion/Type IV pili\n",
      "Flp/Tad pili\n",
      "Flp/Tad pili\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type III secretion system\n",
      "Type IV secretion system\n",
      "Type IV secretion system\n",
      "Type IV secretion system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type IVB secretion/DNA transfer system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VI secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VII secretion system\n",
      "Type VIII secretion system\n",
      "Type VIII secretion system\n",
      "Type VIII secretion system\n",
      "Type VIII secretion system\n",
      "Type VII secretion system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "Type X secretion system\n",
      "Type X secretion system\n",
      "Type X secretion system\n",
      "Type X secretion system\n",
      "Type X secretion system\n",
      "Archaella\n",
      "Type IX secretion/gliding motility system\n",
      "Archaella\n",
      "Type IX secretion/gliding motility system\n",
      "Type IX secretion/gliding motility system\n",
      "nan\n",
      "Archaella\n",
      "nan\n",
      "nan\n"
     ]
    }
   ],
   "source": [
    "for path in cog_def_df['Pathway']:\n",
    "    print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG ID</th>\n",
       "      <th>COG Functional category ID</th>\n",
       "      <th>COG name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Pathway</th>\n",
       "      <th>PubMed ID</th>\n",
       "      <th>PDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4971</th>\n",
       "      <td>COG6081</td>\n",
       "      <td>U</td>\n",
       "      <td>LydA family holin, phage or type X secretion s...</td>\n",
       "      <td>LydA</td>\n",
       "      <td>Type X secretion system</td>\n",
       "      <td>32885520</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4972</th>\n",
       "      <td>COG6082</td>\n",
       "      <td>N</td>\n",
       "      <td>Pre-archaellin peptidase ArlK/FlaK/PibD, inclu...</td>\n",
       "      <td>ArlK</td>\n",
       "      <td>Archaella</td>\n",
       "      <td>25699024;37334237</td>\n",
       "      <td>3S0X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4973</th>\n",
       "      <td>COG6083</td>\n",
       "      <td>U</td>\n",
       "      <td>Type IX system secreted protein, contains C-te...</td>\n",
       "      <td>CTD-A</td>\n",
       "      <td>Type IX secretion/gliding motility system</td>\n",
       "      <td>28396348</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4974</th>\n",
       "      <td>COG6084</td>\n",
       "      <td>KN</td>\n",
       "      <td>Regulator of haloarchaeal motility, adhesion a...</td>\n",
       "      <td>TbsP</td>\n",
       "      <td>Archaella</td>\n",
       "      <td>38204420</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4975</th>\n",
       "      <td>COG6085</td>\n",
       "      <td>U</td>\n",
       "      <td>Type IX system secreted protein, contains C-te...</td>\n",
       "      <td>CTD-B</td>\n",
       "      <td>Type IX secretion/gliding motility system</td>\n",
       "      <td>27005013;28396348</td>\n",
       "      <td>5HFS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4976</th>\n",
       "      <td>COG6086</td>\n",
       "      <td>U</td>\n",
       "      <td>Type IX system secreted protein, contains C-te...</td>\n",
       "      <td>CTD-C</td>\n",
       "      <td>Type IX secretion/gliding motility system</td>\n",
       "      <td>24363341;27005013;28396348</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>COG6088</td>\n",
       "      <td>U</td>\n",
       "      <td>Membrane-anchored component of a predicted arc...</td>\n",
       "      <td>MMP0364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25583072</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4978</th>\n",
       "      <td>COG6090</td>\n",
       "      <td>N</td>\n",
       "      <td>Archaellum component ArlX/FlaX, crenarchaea-sp...</td>\n",
       "      <td>ArlX</td>\n",
       "      <td>Archaella</td>\n",
       "      <td>22081969</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>COG6091</td>\n",
       "      <td>U</td>\n",
       "      <td>Crenarchaeal DNA import system 6TM protein CedA</td>\n",
       "      <td>CedA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26884154</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4980</th>\n",
       "      <td>COG6092</td>\n",
       "      <td>U</td>\n",
       "      <td>Crenarchaeal DNA import system 2TM protein CedA1</td>\n",
       "      <td>CedA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26884154</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       COG ID COG Functional category ID  \\\n",
       "4971  COG6081                          U   \n",
       "4972  COG6082                          N   \n",
       "4973  COG6083                          U   \n",
       "4974  COG6084                         KN   \n",
       "4975  COG6085                          U   \n",
       "4976  COG6086                          U   \n",
       "4977  COG6088                          U   \n",
       "4978  COG6090                          N   \n",
       "4979  COG6091                          U   \n",
       "4980  COG6092                          U   \n",
       "\n",
       "                                               COG name     Gene  \\\n",
       "4971  LydA family holin, phage or type X secretion s...     LydA   \n",
       "4972  Pre-archaellin peptidase ArlK/FlaK/PibD, inclu...     ArlK   \n",
       "4973  Type IX system secreted protein, contains C-te...    CTD-A   \n",
       "4974  Regulator of haloarchaeal motility, adhesion a...     TbsP   \n",
       "4975  Type IX system secreted protein, contains C-te...    CTD-B   \n",
       "4976  Type IX system secreted protein, contains C-te...    CTD-C   \n",
       "4977  Membrane-anchored component of a predicted arc...  MMP0364   \n",
       "4978  Archaellum component ArlX/FlaX, crenarchaea-sp...     ArlX   \n",
       "4979    Crenarchaeal DNA import system 6TM protein CedA     CedA   \n",
       "4980   Crenarchaeal DNA import system 2TM protein CedA1    CedA1   \n",
       "\n",
       "                                        Pathway                   PubMed ID  \\\n",
       "4971                    Type X secretion system                    32885520   \n",
       "4972                                  Archaella           25699024;37334237   \n",
       "4973  Type IX secretion/gliding motility system                    28396348   \n",
       "4974                                  Archaella                    38204420   \n",
       "4975  Type IX secretion/gliding motility system           27005013;28396348   \n",
       "4976  Type IX secretion/gliding motility system  24363341;27005013;28396348   \n",
       "4977                                        NaN                    25583072   \n",
       "4978                                  Archaella                    22081969   \n",
       "4979                                        NaN                    26884154   \n",
       "4980                                        NaN                    26884154   \n",
       "\n",
       "     PDB ID  \n",
       "4971    NaN  \n",
       "4972   3S0X  \n",
       "4973    NaN  \n",
       "4974    NaN  \n",
       "4975   5HFS  \n",
       "4976    NaN  \n",
       "4977    NaN  \n",
       "4978    NaN  \n",
       "4979    NaN  \n",
       "4980    NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_def_df.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4981, 7)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_def_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Heme biosynthesis', 'Arginine biosynthesis', nan,\n",
       "       'Purine salvage', 'Cobalamine/B12 biosynthesis',\n",
       "       'tRNA modification', 'Urea cycle', 'Aminoacyl-tRNA synthetases',\n",
       "       'Proline biosynthesis', 'Purine biosynthesis'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_def_df['Pathway'].unique()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Category: EH\n",
      "Category: EH\n",
      "Category: EF\n",
      "Category: HI\n",
      "Category: EH\n",
      "Category: EG\n",
      "Category: EQ\n",
      "Category: EQ\n",
      "Category: EH\n",
      "Category: OM\n",
      "Category: EH\n",
      "Category: EHJQ\n",
      "Category: KJ\n",
      "Category: EG\n",
      "Category: DP\n",
      "Category: FT\n",
      "Category: JE\n",
      "Category: LJ\n",
      "Category: KT\n",
      "Category: HJ\n",
      "Category: IQ\n",
      "Category: OV\n",
      "Category: TK\n",
      "Category: EM\n",
      "Category: TE\n",
      "Category: FP\n",
      "Category: VI\n",
      "Category: GH\n",
      "Category: FR\n",
      "Category: JHO\n",
      "Category: EP\n",
      "Category: KR\n",
      "Category: DN\n",
      "Category: EF\n",
      "Category: MDT\n",
      "Category: CE\n",
      "Category: GEPR\n",
      "Category: PT\n",
      "Category: ER\n",
      "Category: QR\n",
      "Category: EF\n",
      "Category: EH\n",
      "Category: DH\n",
      "Category: DL\n",
      "Category: FR\n",
      "Category: HC\n",
      "Category: KL\n",
      "Category: HR\n",
      "Category: PT\n",
      "Category: LX\n",
      "Category: HR\n",
      "Category: MU\n",
      "Category: EP\n",
      "Category: CR\n",
      "Category: FT\n",
      "Category: NU\n",
      "Category: CP\n",
      "Category: HC\n",
      "Category: HT\n",
      "Category: HR\n",
      "Category: GER\n",
      "Category: GM\n",
      "Category: FV\n",
      "Category: MU\n",
      "Category: TK\n",
      "Category: OK\n",
      "Category: FV\n",
      "Category: DM\n",
      "Category: FH\n",
      "Category: CO\n",
      "Category: GM\n",
      "Category: MN\n",
      "Category: JLK\n",
      "Category: ET\n",
      "Category: MV\n",
      "Category: CHR\n",
      "Category: LV\n",
      "Category: KL\n",
      "Category: ER\n",
      "Category: DZ\n",
      "Category: MO\n",
      "Category: PH\n",
      "Category: PR\n",
      "Category: EP\n",
      "Category: JV\n",
      "Category: CO\n",
      "Category: CP\n",
      "Category: HI\n",
      "Category: NU\n",
      "Category: KE\n",
      "Category: ER\n",
      "Category: HQ\n",
      "Category: EP\n",
      "Category: MR\n",
      "Category: DN\n",
      "Category: LK\n",
      "Category: JM\n",
      "Category: KT\n",
      "Category: PT\n",
      "Category: FM\n",
      "Category: JO\n",
      "Category: CIR\n",
      "Category: NU\n",
      "Category: CO\n",
      "Category: HR\n",
      "Category: HP\n",
      "Category: KG\n",
      "Category: EG\n",
      "Category: LX\n",
      "Category: ER\n",
      "Category: JR\n",
      "Category: JV\n",
      "Category: TR\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: HO\n",
      "Category: GT\n",
      "Category: NU\n",
      "Category: IR\n",
      "Category: GH\n",
      "Category: CP\n",
      "Category: JO\n",
      "Category: MN\n",
      "Category: TR\n",
      "Category: GT\n",
      "Category: GT\n",
      "Category: NU\n",
      "Category: EQ\n",
      "Category: MR\n",
      "Category: PR\n",
      "Category: DZ\n",
      "Category: KH\n",
      "Category: GH\n",
      "Category: TV\n",
      "Category: UW\n",
      "Category: FL\n",
      "Category: NU\n",
      "Category: MI\n",
      "Category: FR\n",
      "Category: OJ\n",
      "Category: HV\n",
      "Category: TG\n",
      "Category: HE\n",
      "Category: GK\n",
      "Category: FH\n",
      "Category: VT\n",
      "Category: HI\n",
      "Category: CE\n",
      "Category: KT\n",
      "Category: KT\n",
      "Category: FV\n",
      "Category: NU\n",
      "Category: KV\n",
      "Category: GI\n",
      "Category: HJ\n",
      "Category: QV\n",
      "Category: QR\n",
      "Category: FJ\n",
      "Category: DPM\n",
      "Category: HR\n",
      "Category: PQ\n",
      "Category: NU\n",
      "Category: OJ\n",
      "Category: CPO\n",
      "Category: CP\n",
      "Category: TK\n",
      "Category: TK\n",
      "Category: HJ\n",
      "Category: OE\n",
      "Category: FV\n",
      "Category: IR\n",
      "Category: CO\n",
      "Category: DR\n",
      "Category: QR\n",
      "Category: OE\n",
      "Category: FT\n",
      "Category: JT\n",
      "Category: HP\n",
      "Category: AJ\n",
      "Category: KN\n",
      "Category: DI\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: MO\n",
      "Category: JL\n",
      "Category: OT\n",
      "Category: XT\n",
      "Category: PO\n",
      "Category: MG\n",
      "Category: EP\n",
      "Category: CO\n",
      "Category: CP\n",
      "Category: MV\n",
      "Category: NU\n",
      "Category: JV\n",
      "Category: CP\n",
      "Category: MR\n",
      "Category: CO\n",
      "Category: KE\n",
      "Category: NW\n",
      "Category: CP\n",
      "Category: MT\n",
      "Category: CO\n",
      "Category: NT\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: CO\n",
      "Category: ER\n",
      "Category: NW\n",
      "Category: NW\n",
      "Category: FV\n",
      "Category: HV\n",
      "Category: HP\n",
      "Category: CP\n",
      "Category: UW\n",
      "Category: KT\n",
      "Category: KE\n",
      "Category: KX\n",
      "Category: DE\n",
      "Category: CO\n",
      "Category: KN\n",
      "Category: KN\n",
      "Category: NU\n",
      "Category: NW\n",
      "Category: IQ\n",
      "Category: MU\n",
      "Category: MI\n",
      "Category: UX\n",
      "Category: KT\n",
      "Category: GP\n",
      "Category: HI\n",
      "Category: PR\n",
      "Category: TK\n",
      "Category: TK\n",
      "Category: PT\n",
      "Category: JO\n",
      "Category: DM\n",
      "Category: KT\n",
      "Category: KT\n",
      "Category: TU\n",
      "Category: GT\n",
      "Category: QT\n",
      "Category: TK\n",
      "Category: GHR\n",
      "Category: MI\n",
      "Category: JV\n",
      "Category: PR\n",
      "Category: FR\n",
      "Category: OC\n",
      "Category: CO\n",
      "Category: QR\n",
      "Category: HK\n",
      "Category: AJ\n",
      "Category: GM\n",
      "Category: KT\n",
      "Category: CP\n",
      "Category: KT\n",
      "Category: TK\n",
      "Category: TK\n",
      "Category: QC\n",
      "Category: QC\n",
      "Category: MT\n",
      "Category: PR\n",
      "Category: KT\n",
      "Category: GI\n",
      "Category: MX\n",
      "Category: VX\n",
      "Category: JV\n",
      "Category: GM\n",
      "Category: FJ\n",
      "Category: NW\n",
      "Category: KN\n",
      "Category: MI\n",
      "Category: TK\n",
      "Category: MO\n",
      "Category: VT\n",
      "Category: KTN\n",
      "Category: QR\n",
      "Category: GT\n",
      "Category: MR\n",
      "Category: NW\n",
      "Category: NW\n",
      "Category: NW\n",
      "Category: NW\n",
      "Category: NW\n",
      "Category: NU\n",
      "Category: CO\n",
      "Category: CO\n",
      "Category: NW\n",
      "Category: CP\n",
      "Category: OI\n",
      "Category: PU\n",
      "Category: GM\n",
      "Category: DZ\n",
      "Category: CI\n",
      "Category: UW\n",
      "Category: IO\n",
      "Category: QT\n",
      "Category: KV\n",
      "Category: CT\n",
      "Category: KX\n",
      "Category: TU\n",
      "Category: OP\n",
      "Category: CJ\n",
      "Category: CT\n",
      "Category: CT\n",
      "Category: CO\n",
      "Category: CT\n",
      "Category: CO\n",
      "Category: CT\n",
      "Category: CT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DK\n",
      "Category: DK\n",
      "Category: DK\n",
      "Category: DK\n",
      "Category: DT\n",
      "Category: DKO\n",
      "Category: DK\n",
      "Category: DO\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DE\n",
      "Category: DK\n",
      "Category: DT\n",
      "Category: DT\n",
      "Category: DK\n",
      "Category: DL\n",
      "Category: DK\n",
      "Category: DO\n",
      "Category: DT\n",
      "Category: DK\n",
      "Category: DK\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: NU\n",
      "Category: KN\n"
     ]
    }
   ],
   "source": [
    "for i, row in cog_def_df.iterrows():\n",
    "    category = row['COG Functional category ID']\n",
    "    if len(category) > 1:\n",
    "        print(f\"Category: {category}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['K', 'J']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(\"KJ\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.merge(cog_fun_df, cog_def_df, on='COG Functional category ID', how='outer')\n",
    "# doesn't work when the there are more than 1 COG Functional category IDs: DK, DKO, DO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'samplex': ['COG0190',\n",
       "  'COG0826',\n",
       "  'COG1538',\n",
       "  'COG1136',\n",
       "  'COG0577',\n",
       "  'COG0351',\n",
       "  'COG0352',\n",
       "  'COG0422',\n",
       "  'COG0057',\n",
       "  'COG1123',\n",
       "  'COG0359',\n",
       "  'COG0360',\n",
       "  'COG1595',\n",
       "  'COG4775',\n",
       "  'COG0719',\n",
       "  'COG0396',\n",
       "  'COG0782']}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the cog IDs from each sample\n",
    "samplex_cog_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\"\n",
    "\n",
    "sample_name = \"samplex\"\n",
    "cog_ids_dict = {}\n",
    "\n",
    "# Initializing the entry for this sample. The value is a list\n",
    "cog_ids_dict[sample_name] = []\n",
    "\n",
    "with open(samplex_cog_file, mode=\"r\") as fr:\n",
    "    for line in fr:\n",
    "        cog_id = line.strip().split(\":\")[1]\n",
    "        # print(f\"COG ID: {cog_id}\")\n",
    "        \n",
    "        # Appending cog_ids to the sample list value\n",
    "        # It doesn't matter if there are duplicated cog_ids?\n",
    "        cog_ids_dict[sample_name].append(cog_id)\n",
    "        \n",
    "cog_ids_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample, cog_list in cog_ids_dict.items():\n",
    "    for sample_cog_id in cog_list:\n",
    "        if "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG0057 same as COG0057, Functional category: G, Pathway: Glycolysis\n",
      "COG0190 same as COG0190, Functional category: H, Pathway: nan\n",
      "COG0351 same as COG0351, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0352 same as COG0352, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0359 same as COG0359, Functional category: J, Pathway: Ribosome 50S subunit\n",
      "COG0360 same as COG0360, Functional category: J, Pathway: Ribosome 30S subunit\n",
      "COG0396 same as COG0396, Functional category: O, Pathway: nan\n",
      "COG0422 same as COG0422, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0577 same as COG0577, Functional category: V, Pathway: nan\n",
      "COG0719 same as COG0719, Functional category: O, Pathway: nan\n",
      "COG0782 same as COG0782, Functional category: K, Pathway: nan\n",
      "COG0826 same as COG0826, Functional category: J, Pathway: 23S rRNA modification\n",
      "COG1123 same as COG1123, Functional category: O, Pathway: nan\n",
      "COG1136 same as COG1136, Functional category: M, Pathway: nan\n",
      "COG1538 same as COG1538, Functional category: M, Pathway: nan\n",
      "COG1595 same as COG1595, Functional category: K, Pathway: RNA polymerase\n",
      "COG4775 same as COG4775, Functional category: M, Pathway: nan\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'samplex': []}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_fc_dict = {}\n",
    "\n",
    "for i, row in cog_def_df.iterrows():\n",
    "    ncbi_cog_id = row['COG ID']\n",
    "    ncbi_fc = row['COG Functional category ID']\n",
    "    ncbi_pathway = row['Pathway']\n",
    "    \n",
    "    for sample, cog_list in cog_ids_dict.items():\n",
    "        # Initialize dictionary for each new sample\n",
    "        cog_fc_dict[sample] = []\n",
    "        \n",
    "        for sample_cog_id in cog_list:\n",
    "            if sample_cog_id == ncbi_cog_id:\n",
    "                print(f\"{sample_cog_id} same as {ncbi_cog_id}, Functional category: {ncbi_fc}, Pathway: {ncbi_pathway}\")\n",
    "                \n",
    "                cog_fc_dict[sample].append((ncbi_cog_id, ncbi_fc, ncbi_pathway))\n",
    "                \n",
    "cog_fc_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_def_df.loc[cog_def_df['COG ID'] == cog, 'COG Functional category ID'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "category: ['H']\n"
     ]
    }
   ],
   "source": [
    "for i, row in cog_def_df.iterrows():\n",
    "    ncbi_cog_id = row['COG ID']\n",
    "    ncbi_fc = row['COG Functional category ID']\n",
    "    ncbi_pathway = row['Pathway']\n",
    "    \n",
    "    if ncbi_cog_id == 'COG0190':\n",
    "        category = cog_def_df.loc[cog_def_df['COG ID'] == 'COG0190', 'COG Functional category ID'].values\n",
    "        print(f\"category: {category}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG0190 same as COG0190, Functional category: H, Pathway: nan\n",
      "COG0826 same as COG0826, Functional category: J, Pathway: 23S rRNA modification\n",
      "COG1538 same as COG1538, Functional category: M, Pathway: nan\n",
      "COG1136 same as COG1136, Functional category: M, Pathway: nan\n",
      "COG0577 same as COG0577, Functional category: V, Pathway: nan\n",
      "COG0351 same as COG0351, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0352 same as COG0352, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0422 same as COG0422, Functional category: H, Pathway: Thiamine biosynthesis\n",
      "COG0057 same as COG0057, Functional category: G, Pathway: Glycolysis\n",
      "COG1123 same as COG1123, Functional category: O, Pathway: nan\n",
      "COG0359 same as COG0359, Functional category: J, Pathway: Ribosome 50S subunit\n",
      "COG0360 same as COG0360, Functional category: J, Pathway: Ribosome 30S subunit\n",
      "COG1595 same as COG1595, Functional category: K, Pathway: RNA polymerase\n",
      "COG4775 same as COG4775, Functional category: M, Pathway: nan\n",
      "COG0719 same as COG0719, Functional category: O, Pathway: nan\n",
      "COG0396 same as COG0396, Functional category: O, Pathway: nan\n",
      "COG0782 same as COG0782, Functional category: K, Pathway: nan\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'samplex': [('COG0190', 'H', nan),\n",
       "  ('COG0826', 'J', '23S rRNA modification'),\n",
       "  ('COG1538', 'M', nan),\n",
       "  ('COG1136', 'M', nan),\n",
       "  ('COG0577', 'V', nan),\n",
       "  ('COG0351', 'H', 'Thiamine biosynthesis'),\n",
       "  ('COG0352', 'H', 'Thiamine biosynthesis'),\n",
       "  ('COG0422', 'H', 'Thiamine biosynthesis'),\n",
       "  ('COG0057', 'G', 'Glycolysis'),\n",
       "  ('COG1123', 'O', nan),\n",
       "  ('COG0359', 'J', 'Ribosome 50S subunit'),\n",
       "  ('COG0360', 'J', 'Ribosome 30S subunit'),\n",
       "  ('COG1595', 'K', 'RNA polymerase'),\n",
       "  ('COG4775', 'M', nan),\n",
       "  ('COG0719', 'O', nan),\n",
       "  ('COG0396', 'O', nan),\n",
       "  ('COG0782', 'K', nan)]}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample, cog_list in cog_ids_dict.items():\n",
    "    # Initialize dictionary for each new sample\n",
    "    cog_fc_dict[sample] = []\n",
    "    \n",
    "    for sample_cog_id in cog_list:\n",
    "        for i, row in cog_def_df.iterrows():\n",
    "            ncbi_cog_id = row['COG ID']\n",
    "            ncbi_fc = row['COG Functional category ID']\n",
    "            ncbi_pathway = row['Pathway']\n",
    "            if sample_cog_id == ncbi_cog_id:\n",
    "                print(f\"{sample_cog_id} same as {ncbi_cog_id}, Functional category: {ncbi_fc}, Pathway: {ncbi_pathway}\")\n",
    "                \n",
    "                cog_fc_dict[sample].append((ncbi_cog_id, ncbi_fc, ncbi_pathway))\n",
    "cog_fc_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG ID: COG0190, category: H, Pathway: nan\n",
      "category: ['H']\n"
     ]
    }
   ],
   "source": [
    "for sample, cog_list in cog_ids_dict.items():\n",
    "    # Initialize dictionary for each new sample\n",
    "    cog_fc_dict[sample] = []\n",
    "    \n",
    "    for sample_cog_id in cog_list:\n",
    "        for i, row in cog_def_df.iterrows():\n",
    "            ncbi_cog_id = row['COG ID']\n",
    "            ncbi_fc = row['COG Functional category ID']\n",
    "            ncbi_pathway = row['Pathway']\n",
    "            if sample_cog_id == 'COG0190' and ncbi_cog_id == 'COG0190':\n",
    "                print(f\"COG ID: {sample_cog_id}, category: {ncbi_fc}, Pathway: {ncbi_pathway}\")\n",
    "                category = cog_def_df.loc[cog_def_df['COG ID'] == sample_cog_id, 'COG Functional category ID'].values\n",
    "                print(f\"category: {category}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samplex\n"
     ]
    }
   ],
   "source": [
    "samples_cog_dir = Path(\"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples\")\n",
    "\n",
    "for item in samples_cog_dir.iterdir():\n",
    "    if item.is_file():\n",
    "        sample = item.stem.split(\"_\")[0]\n",
    "        print(sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# P.3. Script 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description  samplex  \n",
       "0             Cell wall/membrane/envelope biogenesis      3.0  \n",
       "1                                 Defense mechanisms      1.0  \n",
       "2  Posttranslational modification, protein turnov...      3.0  \n",
       "3    Translation, ribosomal structure and biogenesis      2.5  \n",
       "4                                      Transcription      2.0  \n",
       "5                Amino acid transport and metabolism      0.5  \n",
       "6                  Coenzyme transport and metabolism      3.0  \n",
       "7              Carbohydrate transport and metabolism      1.0  \n",
       "8                   General function prediction only      1.0  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "# Load cog-24.def.tab file (COG ID to Functional Category mapping)\n",
    "def load_cog_def(cog_def_file):\n",
    "    cog_def_df = pd.read_csv(\n",
    "        cog_def_file, \n",
    "        sep='\\t', \n",
    "        names=['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway', 'PubMed ID', 'PDB ID']\n",
    "        )\n",
    "    cog_def_df['COG ID'] = cog_def_df['COG ID'].str.strip()  # Strip any spaces\n",
    "    return cog_def_df\n",
    "\n",
    "# Load cog-24.fun.edited.tab file (Functional Category descriptions and groups)\n",
    "def load_cog_fun(cog_fun_file):\n",
    "    cog_fun_df = pd.read_csv(\n",
    "        cog_fun_file, \n",
    "        sep='\\t', \n",
    "        names=['COG Functional category ID', 'Functional group', 'RGB color', 'FC description']\n",
    "        )\n",
    "    return cog_fun_df\n",
    "\n",
    "# Load sample-specific protein-COG mappings\n",
    "def load_sample_protein_cog(sample_protein_cog_file):\n",
    "    sample_cog_df = pd.read_csv(sample_protein_cog_file, sep='\\t', header=None, names=['Protein', 'COG'])\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.strip()\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.replace('COG:', '')  # Clean up COG prefixes\n",
    "    return sample_cog_df\n",
    "\n",
    "# Handle combined functional categories (like EHJQ, etc.)\n",
    "def map_combined_categories(cog_categories):\n",
    "    return list(cog_categories)  # Split the string into individual categories (like EHJQ -> ['E', 'H', 'J', 'Q'])\n",
    "\n",
    "# Function to summarize the functional categories for a given sample\n",
    "def summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df):\n",
    "    category_counts = defaultdict(int)  # Count occurrences of each functional category\n",
    "    pathway_counts = defaultdict(int) # Count occurrences of each Pathway\n",
    "    \n",
    "    for cog in sample_cog_df['COG']:\n",
    "        # Find the functional category/ies for the COG\n",
    "        # print(f\"Processing COG: {cog}\")\n",
    "        functional_categories = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'COG Functional category ID'].values\n",
    "        pathways = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'Pathway'].values[0]\n",
    "        \n",
    "        # Handle case when no category is found\n",
    "        if len(functional_categories) == 0:\n",
    "            print(f\"COG ID {cog} not found in cog_def_df.\")\n",
    "            continue\n",
    "        \n",
    "        # print(f\"Functional categories found: {functional_categories}\")\n",
    "        functional_categories = functional_categories[0]  # Extract the category\n",
    "        # print(f\"Functional categories found: {map_combined_categories(functional_categories)}\")\n",
    "        # print(f\"Number of Functional categories found: {len(map_combined_categories(functional_categories))}\")\n",
    "        \n",
    "        # Get the number of categories for this COG\n",
    "        num_categories = len(map_combined_categories(functional_categories))\n",
    "        # Handle combined categories (split them into individual letters, e.g., 'ER' -> ['E', 'R'])\n",
    "        for category in map_combined_categories(functional_categories):\n",
    "            # Add 1 / num_categories to each category\n",
    "            category_counts[category] += 1 / num_categories\n",
    "\n",
    "    # Collect rows for the summary DataFrame\n",
    "    summary_rows = []\n",
    "    \n",
    "    for category, count in category_counts.items():\n",
    "        group = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'Functional group'].values[0]\n",
    "        description = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'FC description'].values[0]\n",
    "        color = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'RGB color'].values[0]\n",
    "        # Collect the row as a dictionary\n",
    "        summary_rows.append(\n",
    "            {\n",
    "                'Functional category': category,\n",
    "                'Functional group': group,\n",
    "                'Count': count,\n",
    "                'FC description': description,\n",
    "                'Color': color\n",
    "            }\n",
    "                            )\n",
    "\n",
    "    # create the summary DataFrame with category, group, and count columns\n",
    "    summary_df = pd.DataFrame(summary_rows)\n",
    "    \n",
    "    return summary_df\n",
    "\n",
    "# Main function to process each sample and summarize the functional categories\n",
    "def process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file):\n",
    "    # Load the data\n",
    "    cog_def_df = load_cog_def(cog_def_file)\n",
    "    cog_fun_df = load_cog_fun(cog_fun_file)\n",
    "    sample_cog_df = load_sample_protein_cog(sample_protein_cog_file)\n",
    "\n",
    "    # Summarize the functional categories\n",
    "    summary_df = summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df)\n",
    "    # Dataframe for comparison between samples\n",
    "    summary_df['Functional group'] = summary_df['Functional group'].replace(\n",
    "        {\n",
    "            1: 'Information storage and processing',\n",
    "            2: 'Cellular processes and signaling',\n",
    "            3: 'Metabolism',\n",
    "            4: 'Poorly characterized',\n",
    "        }\n",
    "            )\n",
    "    # Group the rows with the same 'Functional group'\n",
    "    summary_df = summary_df.sort_values(by='Functional group').reset_index(drop=True)\n",
    "    # Create non-normalized DataFrame\n",
    "    non_normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    non_normalized_df = non_normalized_df.rename(columns={'Count': sample})\n",
    "    # Create normalized DataFrame\n",
    "    ## Normalize the counts to relative abundance\n",
    "    normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    normalized_df[sample] = normalized_df['Count'] / normalized_df['Count'].sum()\n",
    "    normalized_df.drop(columns=['Count'], inplace=True)\n",
    "    # Return DataFrames\n",
    "    return non_normalized_df, normalized_df\n",
    "\n",
    "# Example usage:\n",
    "sample = \"samplex\"\n",
    "# sample_protein_cog_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\"\n",
    "sample_protein_cog_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples/sampley_protein-cog.tsv\"\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "non_normalized_df, normalized_df = process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file)\n",
    "non_normalized_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>0.147059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>0.117647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description   samplex  \n",
       "0             Cell wall/membrane/envelope biogenesis  0.176471  \n",
       "1                                 Defense mechanisms  0.058824  \n",
       "2  Posttranslational modification, protein turnov...  0.176471  \n",
       "3    Translation, ribosomal structure and biogenesis  0.147059  \n",
       "4                                      Transcription  0.117647  \n",
       "5                Amino acid transport and metabolism  0.029412  \n",
       "6                  Coenzyme transport and metabolism  0.176471  \n",
       "7              Carbohydrate transport and metabolism  0.058824  \n",
       "8                   General function prediction only  0.058824  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing samplex from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\n",
      "Processing sampley from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/sampley_protein-cog.tsv\n"
     ]
    }
   ],
   "source": [
    "samples_cog_dir = Path(\"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples\")\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "\n",
    "# Initialize lists to hold DataFrames\n",
    "non_normalized_list = []\n",
    "normalized_list = []\n",
    "\n",
    "# Iterate over files in the sample directory\n",
    "for item in samples_cog_dir.iterdir():\n",
    "    if item.is_file() and item.suffix == \".tsv\":  # Filter for .tsv files\n",
    "        sample_protein_cog_file = item\n",
    "        # Extract sample name from filename\n",
    "        sample = item.stem.split(\"_\")[0]\n",
    "        print(f\"Processing {sample} from file: {item}\")\n",
    "        \n",
    "        # Process the sample\n",
    "        non_normalized_df, normalized_df = process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file)\n",
    "        # Append the results to the list\n",
    "        non_normalized_list.append(non_normalized_df)\n",
    "        normalized_list.append(normalized_df)\n",
    "        \n",
    "# Merge all DataFrames by 'Functional group' and 'FC description'\n",
    "non_normalized_combined = pd.concat(non_normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()\n",
    "normalized_combined = pd.concat(normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description  samplex  sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis      3.0      3.0  \n",
       "1                                 Defense mechanisms      1.0      1.0  \n",
       "2  Posttranslational modification, protein turnov...      3.0      3.0  \n",
       "3                                      Transcription      2.0      2.0  \n",
       "4    Translation, ribosomal structure and biogenesis      3.0      2.5  \n",
       "5                Amino acid transport and metabolism      0.0      0.5  \n",
       "6              Carbohydrate transport and metabolism      1.0      1.0  \n",
       "7                  Coenzyme transport and metabolism      4.0      3.0  \n",
       "8                   General function prediction only      0.0      1.0  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.117647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.147059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description   samplex   sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis  0.176471  0.176471  \n",
       "1                                 Defense mechanisms  0.058824  0.058824  \n",
       "2  Posttranslational modification, protein turnov...  0.176471  0.176471  \n",
       "3                                      Transcription  0.117647  0.117647  \n",
       "4    Translation, ribosomal structure and biogenesis  0.176471  0.147059  \n",
       "5                Amino acid transport and metabolism  0.000000  0.029412  \n",
       "6              Carbohydrate transport and metabolism  0.058824  0.058824  \n",
       "7                  Coenzyme transport and metabolism  0.235294  0.176471  \n",
       "8                   General function prediction only  0.000000  0.058824  "
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_combined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Key Improvements:**\n",
    "\n",
    "1. **Filter for `.tsv` Files**: \n",
    "   - The check `item.suffix == \".tsv\"` ensures that only `.tsv` files are processed, in case there are other types of files in the directory.\n",
    "   \n",
    "2. **Use `pd.concat()` Instead of `append()`**:\n",
    "   - Instead of appending DataFrames to a list and then merging later, I’ve used `pd.concat()` with a **groupby** to merge the DataFrames on the fly.\n",
    "   - The `groupby(['Functional group', 'FC description'])` ensures that if the same functional group and description appear in multiple samples, they are aggregated (you can sum the counts for each group).\n",
    "   \n",
    "3. **Summing the Data**: \n",
    "   - Using `.groupby(...).sum()` allows the counts across different samples to be aggregated into a single DataFrame.\n",
    "\n",
    "**What Happens in the `process_all_samples` Function:**\n",
    "\n",
    "- **Iterates over each file**: For each file in the directory, it processes the sample using `process_sample`.\n",
    "- **Appends the results**: The non-normalized and normalized DataFrames are appended to their respective lists.\n",
    "- **Merges the DataFrames**: At the end, the DataFrames are merged by `'Functional group'` and `'FC description'` using `pd.concat()` and `groupby`.\n",
    "\n",
    "**Benefits**:\n",
    "\n",
    "- **More Pythonic**: Using `pd.concat()` and filtering the file types makes the code cleaner and avoids the need for post-processing steps.\n",
    "- **Efficient**: Merging on-the-fly reduces memory overhead and avoids multiple passes through the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_all_samples(samples_cog_dir, cog_def_file, cog_fun_file):\n",
    "    # Initialize lists to hold DataFrames\n",
    "    non_normalized_list = []\n",
    "    normalized_list = []\n",
    "    \n",
    "    # Iterate over files in the sample directory\n",
    "    for item in samples_cog_dir.iterdir():\n",
    "        if item.is_file() and item.suffix == \".tsv\":  # Filter for .tsv files\n",
    "            sample_protein_cog_file = item\n",
    "            sample = item.stem.split(\"_\")[0]  # Extract sample name from filename\n",
    "            print(f\"Processing {sample} from file: {item}\")\n",
    "            \n",
    "            # Process the sample\n",
    "            non_normalized_df, normalized_df = process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file)\n",
    "            \n",
    "            # Append the results to the list\n",
    "            non_normalized_list.append(non_normalized_df)\n",
    "            normalized_list.append(normalized_df)\n",
    "    \n",
    "    # Concatenate and merge all DataFrames by 'Functional group' and 'FC description'\n",
    "    non_normalized_combined = pd.concat(non_normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()\n",
    "    normalized_combined = pd.concat(normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()\n",
    "    \n",
    "    return non_normalized_combined, normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing samplex from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\n",
      "Processing sampley from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/sampley_protein-cog.tsv\n"
     ]
    }
   ],
   "source": [
    "samples_cog_dir = Path(\"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples\")\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "\n",
    "non_normalized_combined, normalized_combined = process_all_samples(samples_cog_dir, cog_def_file, cog_fun_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description  samplex  sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis      3.0      3.0  \n",
       "1                                 Defense mechanisms      1.0      1.0  \n",
       "2  Posttranslational modification, protein turnov...      3.0      3.0  \n",
       "3                                      Transcription      2.0      2.0  \n",
       "4    Translation, ribosomal structure and biogenesis      3.0      2.5  \n",
       "5                Amino acid transport and metabolism      0.0      0.5  \n",
       "6              Carbohydrate transport and metabolism      1.0      1.0  \n",
       "7                  Coenzyme transport and metabolism      4.0      3.0  \n",
       "8                   General function prediction only      0.0      1.0  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.117647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.147059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description   samplex   sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis  0.176471  0.176471  \n",
       "1                                 Defense mechanisms  0.058824  0.058824  \n",
       "2  Posttranslational modification, protein turnov...  0.176471  0.176471  \n",
       "3                                      Transcription  0.117647  0.117647  \n",
       "4    Translation, ribosomal structure and biogenesis  0.176471  0.147059  \n",
       "5                Amino acid transport and metabolism  0.000000  0.029412  \n",
       "6              Carbohydrate transport and metabolism  0.058824  0.058824  \n",
       "7                  Coenzyme transport and metabolism  0.235294  0.176471  \n",
       "8                   General function prediction only  0.000000  0.058824  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway',\n",
      "       'PubMed ID', 'PDB ID'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(cog_def_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column names in cog_def_df: Index(['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway',\n",
      "       'PubMed ID', 'PDB ID'],\n",
      "      dtype='object')\n",
      "category: ['H']\n"
     ]
    }
   ],
   "source": [
    "# Check and clean column names\n",
    "print(\"Column names in cog_def_df:\", cog_def_df.columns)\n",
    "\n",
    "# Ensure column names are stripped of any leading/trailing spaces\n",
    "cog_def_df.columns = cog_def_df.columns.str.strip()\n",
    "\n",
    "# Try fetching the functional category again\n",
    "functional_categories = cog_def_df.loc[cog_def_df['COG ID'] == 'COG0190', 'COG Functional category ID'].values\n",
    "print(f\"category: {functional_categories}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# P.3.5. Script 1.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "# Load cog-24.def.tab file (COG ID to Functional Category mapping)\n",
    "def load_cog_def(cog_def_file):\n",
    "    cog_def_df = pd.read_csv(\n",
    "        cog_def_file, \n",
    "        sep='\\t', \n",
    "        names=['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway', 'PubMed ID', 'PDB ID']\n",
    "        )\n",
    "    cog_def_df['COG ID'] = cog_def_df['COG ID'].str.strip()  # Strip any spaces\n",
    "    return cog_def_df\n",
    "\n",
    "# Load cog-24.fun.edited.tab file (Functional Category descriptions and groups)\n",
    "def load_cog_fun(cog_fun_file):\n",
    "    cog_fun_df = pd.read_csv(\n",
    "        cog_fun_file, \n",
    "        sep='\\t', \n",
    "        names=['COG Functional category ID', 'Functional group', 'RGB color', 'FC description']\n",
    "        )\n",
    "    return cog_fun_df\n",
    "\n",
    "# Load sample-specific protein-COG mappings\n",
    "def load_sample_protein_cog(sample_protein_cog_file):\n",
    "    sample_cog_df = pd.read_csv(sample_protein_cog_file, sep='\\t', header=None, names=['Protein', 'COG'])\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.strip()\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.replace('COG:', '')  # Clean up COG prefixes\n",
    "    return sample_cog_df\n",
    "\n",
    "# Handle combined functional categories (like EHJQ, etc.)\n",
    "def map_combined_categories(cog_categories):\n",
    "    return list(cog_categories)  # Split the string into individual categories (like EHJQ -> ['E', 'H', 'J', 'Q'])\n",
    "\n",
    "# Function to summarize the functional categories for a given sample\n",
    "def summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df):\n",
    "    category_counts = defaultdict(int)  # Count occurrences of each functional category\n",
    "    pathway_counts = defaultdict(int) # Count occurrences of each Pathway\n",
    "    \n",
    "    for cog in sample_cog_df['COG']:\n",
    "        # Find the functional category/ies for the COG\n",
    "        # print(f\"Processing COG: {cog}\")\n",
    "        functional_categories = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'COG Functional category ID'].values\n",
    "        pathways = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'Pathway'].values[0]\n",
    "        \n",
    "        # Handle case when no category is found\n",
    "        if len(functional_categories) == 0:\n",
    "            print(f\"COG ID {cog} not found in cog_def_df.\")\n",
    "            continue\n",
    "        \n",
    "        # print(f\"Functional categories found: {functional_categories}\")\n",
    "        functional_categories = functional_categories[0]  # Extract the category\n",
    "        # print(f\"Functional categories found: {map_combined_categories(functional_categories)}\")\n",
    "        # print(f\"Number of Functional categories found: {len(map_combined_categories(functional_categories))}\")\n",
    "        \n",
    "        # Get the number of categories for this COG\n",
    "        num_categories = len(map_combined_categories(functional_categories))\n",
    "        # Handle combined categories (split them into individual letters, e.g., 'ER' -> ['E', 'R'])\n",
    "        for category in map_combined_categories(functional_categories):\n",
    "            # Add 1 / num_categories to each category\n",
    "            category_counts[category] += 1 / num_categories\n",
    "\n",
    "    # Collect rows for the summary DataFrame\n",
    "    summary_rows = []\n",
    "    \n",
    "    for category, count in category_counts.items():\n",
    "        group = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'Functional group'].values[0]\n",
    "        description = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'FC description'].values[0]\n",
    "        color = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'RGB color'].values[0]\n",
    "        # Collect the row as a dictionary\n",
    "        summary_rows.append(\n",
    "            {\n",
    "                'Functional category': category,\n",
    "                'Functional group': group,\n",
    "                'Count': count,\n",
    "                'FC description': description,\n",
    "                'Color': color\n",
    "            }\n",
    "                            )\n",
    "\n",
    "    # create the summary DataFrame with category, group, and count columns\n",
    "    summary_df = pd.DataFrame(summary_rows)\n",
    "    \n",
    "    return summary_df\n",
    "\n",
    "# Main function to process each sample and summarize the functional categories\n",
    "def process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file):\n",
    "    # Load the data\n",
    "    cog_def_df = load_cog_def(cog_def_file)\n",
    "    cog_fun_df = load_cog_fun(cog_fun_file)\n",
    "    sample_cog_df = load_sample_protein_cog(sample_protein_cog_file)\n",
    "\n",
    "    # Summarize the functional categories\n",
    "    summary_df = summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df)\n",
    "    # Dataframe for comparison between samples\n",
    "    summary_df['Functional group'] = summary_df['Functional group'].replace(\n",
    "        {\n",
    "            1: 'Information storage and processing',\n",
    "            2: 'Cellular processes and signaling',\n",
    "            3: 'Metabolism',\n",
    "            4: 'Poorly characterized',\n",
    "        }\n",
    "            )\n",
    "    # Group the rows with the same 'Functional group'\n",
    "    summary_df = summary_df.sort_values(by='Functional group').reset_index(drop=True)\n",
    "    # Create non-normalized DataFrame\n",
    "    non_normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    non_normalized_df = non_normalized_df.rename(columns={'Count': sample})\n",
    "    # Create normalized DataFrame\n",
    "    ## Normalize the counts to relative abundance\n",
    "    normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    normalized_df[sample] = normalized_df['Count'] / normalized_df['Count'].sum()\n",
    "    normalized_df.drop(columns=['Count'], inplace=True)\n",
    "    # Return DataFrames\n",
    "    return non_normalized_df, normalized_df\n",
    "\n",
    "def process_all_samples(samples_cog_dir, output_dir, cog_def_file, cog_fun_file):\n",
    "    # Initialize lists to hold DataFrames\n",
    "    non_normalized_list = []\n",
    "    normalized_list = []\n",
    "    \n",
    "    # Iterate over files in the sample directory\n",
    "    for item in samples_cog_dir.iterdir():\n",
    "        if item.is_file() and item.suffix == \".tsv\":  # Filter for .tsv files\n",
    "            sample_protein_cog_file = item\n",
    "            sample = item.stem.split(\"_\")[0]  # Extract sample name from filename\n",
    "            print(f\"Processing {sample} from file: {item}\")\n",
    "            \n",
    "            # Process the sample\n",
    "            non_normalized_df, normalized_df = process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file)\n",
    "            \n",
    "            # Append the results to the list\n",
    "            non_normalized_list.append(non_normalized_df)\n",
    "            normalized_list.append(normalized_df)\n",
    "    \n",
    "    # Concatenate and merge all DataFrames by 'Functional group' and 'FC description'\n",
    "    non_normalized_combined = pd.concat(non_normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()\n",
    "    normalized_combined = pd.concat(normalized_list).groupby(['Functional group', 'FC description'], as_index=False).sum()\n",
    "    \n",
    "    # save\n",
    "    non_normalized_combined.to_csv(output_dir / \"non_normalized.csv\", index=False)\n",
    "    normalized_combined.to_csv(output_dir / \"normalized.csv\", index=False)\n",
    "    print(f\"\\nNon normalized CSV saved to: {output_dir / 'non_normalized.csv'}\")\n",
    "    print(f\"Normalized CSV saved to: {output_dir / 'normalized.csv'}\")\n",
    "    \n",
    "    return non_normalized_combined, normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing samplex from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\n",
      "Processing sampley from file: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples/sampley_protein-cog.tsv\n",
      "\n",
      "Non normalized CSV saved to: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples_fp/non_normalized.csv\n",
      "Normalized CSV saved to: /home/koala/my_projects/metagenome_metabolic_profiling/data/samples_fp/normalized.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description  samplex  sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis      3.0      3.0  \n",
       "1                                 Defense mechanisms      1.0      1.0  \n",
       "2  Posttranslational modification, protein turnov...      3.0      3.0  \n",
       "3                                      Transcription      2.0      2.0  \n",
       "4    Translation, ribosomal structure and biogenesis      3.0      2.5  \n",
       "5                Amino acid transport and metabolism      0.0      0.5  \n",
       "6              Carbohydrate transport and metabolism      1.0      1.0  \n",
       "7                  Coenzyme transport and metabolism      4.0      3.0  \n",
       "8                   General function prediction only      0.0      1.0  "
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_cog_dir = Path(\"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples\")\n",
    "output_dir = Path(\"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples_fp\")\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "\n",
    "non_normalized_combined, normalized_combined = process_all_samples(samples_cog_dir, output_dir, cog_def_file, cog_fun_file)\n",
    "\n",
    "non_normalized_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Functional group</th>\n",
       "      <th>FC description</th>\n",
       "      <th>samplex</th>\n",
       "      <th>sampley</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cellular processes and signaling</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.117647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Information storage and processing</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.147059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Poorly characterized</td>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Functional group  \\\n",
       "0    Cellular processes and signaling   \n",
       "1    Cellular processes and signaling   \n",
       "2    Cellular processes and signaling   \n",
       "3  Information storage and processing   \n",
       "4  Information storage and processing   \n",
       "5                          Metabolism   \n",
       "6                          Metabolism   \n",
       "7                          Metabolism   \n",
       "8                Poorly characterized   \n",
       "\n",
       "                                      FC description   samplex   sampley  \n",
       "0             Cell wall/membrane/envelope biogenesis  0.176471  0.176471  \n",
       "1                                 Defense mechanisms  0.058824  0.058824  \n",
       "2  Posttranslational modification, protein turnov...  0.176471  0.176471  \n",
       "3                                      Transcription  0.117647  0.117647  \n",
       "4    Translation, ribosomal structure and biogenesis  0.176471  0.147059  \n",
       "5                Amino acid transport and metabolism  0.000000  0.029412  \n",
       "6              Carbohydrate transport and metabolism  0.058824  0.058824  \n",
       "7                  Coenzyme transport and metabolism  0.235294  0.176471  \n",
       "8                   General function prediction only  0.000000  0.058824  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalized_combined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# P.4. Script 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "\n",
    "# Load cog-24.def.tab file (COG ID to Functional Category mapping)\n",
    "def load_cog_def(cog_def_file):\n",
    "    cog_def_df = pd.read_csv(\n",
    "        cog_def_file, \n",
    "        sep='\\t', \n",
    "        names=['COG ID', 'COG Functional category ID', 'COG name', 'Gene', 'Pathway', 'PubMed ID', 'PDB ID']\n",
    "        )\n",
    "    cog_def_df['COG ID'] = cog_def_df['COG ID'].str.strip()  # Strip any spaces\n",
    "    return cog_def_df\n",
    "\n",
    "# Load cog-24.fun.edited.tab file (Functional Category descriptions and groups)\n",
    "def load_cog_fun(cog_fun_file):\n",
    "    cog_fun_df = pd.read_csv(\n",
    "        cog_fun_file, \n",
    "        sep='\\t', \n",
    "        names=['COG Functional category ID', 'Functional group', 'RGB color', 'FC description']\n",
    "        )\n",
    "    return cog_fun_df\n",
    "\n",
    "# Load sample-specific protein-COG mappings\n",
    "def load_sample_protein_cog(sample_protein_cog_file):\n",
    "    sample_cog_df = pd.read_csv(sample_protein_cog_file, sep='\\t', header=None, names=['Protein', 'COG'])\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.strip()\n",
    "    sample_cog_df['COG'] = sample_cog_df['COG'].str.replace('COG:', '')  # Clean up COG prefixes\n",
    "    return sample_cog_df\n",
    "\n",
    "# Handle combined functional categories (like EHJQ, etc.)\n",
    "def map_combined_categories(cog_categories):\n",
    "    return list(cog_categories)  # Split the string into individual categories (like EHJQ -> ['E', 'H', 'J', 'Q'])\n",
    "\n",
    "# Function to summarize the functional categories for a given sample\n",
    "def summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df):\n",
    "    category_counts = defaultdict(int)  # Count occurrences of each functional category\n",
    "    pathway_counts = defaultdict(int) # Count occurrences of each Pathway\n",
    "    \n",
    "    for cog in sample_cog_df['COG']:\n",
    "        # Find the functional category/ies for the COG\n",
    "        print(f\"Processing COG: {cog}\")\n",
    "        functional_categories = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'COG Functional category ID'].values\n",
    "        pathways = cog_def_df.loc[cog_def_df['COG ID'] == cog, 'Pathway'].values[0]\n",
    "        \n",
    "        # Handle case when no category is found\n",
    "        if len(functional_categories) == 0:\n",
    "            print(f\"COG ID {cog} not found in cog_def_df.\")\n",
    "            continue\n",
    "        \n",
    "        # print(f\"Functional categories found: {functional_categories}\")\n",
    "        functional_categories = functional_categories[0]  # Extract the category\n",
    "        # print(f\"Functional categories found: {map_combined_categories(functional_categories)}\")\n",
    "        # print(f\"Number of Functional categories found: {len(map_combined_categories(functional_categories))}\")\n",
    "        \n",
    "        # Get the number of categories for this COG\n",
    "        num_categories = len(map_combined_categories(functional_categories))\n",
    "        # Handle combined categories (split them into individual letters, e.g., 'ER' -> ['E', 'R'])\n",
    "        for category in map_combined_categories(functional_categories):\n",
    "            # Add 1 / num_categories to each category\n",
    "            category_counts[category] += 1 / num_categories\n",
    "\n",
    "    # Collect rows for the summary DataFrame\n",
    "    summary_rows = []\n",
    "    \n",
    "    for category, count in category_counts.items():\n",
    "        group = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'Functional group'].values[0]\n",
    "        description = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'FC description'].values[0]\n",
    "        color = cog_fun_df.loc[cog_fun_df['COG Functional category ID'] == category, 'RGB color'].values[0]\n",
    "        # Collect the row as a dictionary\n",
    "        summary_rows.append(\n",
    "            {\n",
    "                'Functional category': category,\n",
    "                'Functional group': group,\n",
    "                'Count': count,\n",
    "                'FC description': description,\n",
    "                'Color': color\n",
    "            }\n",
    "                            )\n",
    "\n",
    "    # create the summary DataFrame with category, group, and count columns\n",
    "    summary_df = pd.DataFrame(summary_rows)\n",
    "    \n",
    "    return summary_df\n",
    "\n",
    "# Main function to process each sample and summarize the functional categories\n",
    "def process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file):\n",
    "    # Load the data\n",
    "    cog_def_df = load_cog_def(cog_def_file)\n",
    "    cog_fun_df = load_cog_fun(cog_fun_file)\n",
    "    sample_cog_df = load_sample_protein_cog(sample_protein_cog_file)\n",
    "\n",
    "    # Summarize the functional categories\n",
    "    summary_df = summarize_functional_categories(sample_cog_df, cog_def_df, cog_fun_df)\n",
    "    # Dataframe for comparison between samples\n",
    "    summary_df['Functional group'] = summary_df['Functional group'].replace(\n",
    "        {\n",
    "            1: 'Information storage and processing',\n",
    "            2: 'Cellular processes and signaling',\n",
    "            3: 'Metabolism',\n",
    "            4: 'Poorly characterized',\n",
    "        }\n",
    "            )\n",
    "    # Group the rows with the same 'Functional group'\n",
    "    summary_df = summary_df.sort_values(by='Functional group').reset_index(drop=True)\n",
    "    # Create non-normalized DataFrame\n",
    "    non_normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    non_normalized_df = non_normalized_df.rename(columns={'Count': sample})\n",
    "    # Create normalized DataFrame\n",
    "    ## Normalize the counts to relative abundance\n",
    "    normalized_df = summary_df[['Functional group', 'FC description', 'Count']].copy()\n",
    "    normalized_df[sample] = normalized_df['Count'] / normalized_df['Count'].sum()\n",
    "    normalized_df.drop(columns=['Count'], inplace=True)\n",
    "    # Return DataFrames\n",
    "    return non_normalized_df, normalized_df\n",
    "\n",
    "# Example usage:\n",
    "sample = \"samplex\"\n",
    "# sample_protein_cog_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples/samplex_protein-cog.tsv\"\n",
    "sample_protein_cog_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/samples/sampley_protein-cog.tsv\"\n",
    "cog_def_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.def.tab\"\n",
    "cog_fun_file = \"/home/koala/my_projects/metagenome_metabolic_profiling/data/cog-24.fun.edited.tab\"\n",
    "non_normalized_df, normalized_df = process_sample(sample, sample_protein_cog_file, cog_def_file, cog_fun_file)\n",
    "non_normalized_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
